### PR TITLE
#2383 - Horizontal Autoscaler from replication controller to deployment config

### PR DESCRIPTION
**Pull request by guru-aot**
- [x] Horizontal autoscaler target needs to be changed from Replication controller to deployment config.
- [x] Apply the Horizontal autoscaler for the pods, change the deploy.yaml and apply them during the deployments.

Without HPA not properly assigned
image

With HPA properly assigned
image
image

Formio Load testing
WIth 2 pods it scaled up to 6 pods for 400 Vu
image
image
image

With 6 pods running the same sets of load for 400 Vu
image
image
image

With 6 pods running 300 Vu
image
image
image

With 6 pods running 200 Vu
image
image
image

With 6 pods running 100 Vu
image
image
image

With 2 pods running 100 Vu it started the 3rd pod
image
image
image

### CODE REVIEW

File: devops/openshift/api-deploy.yml [hunk 1]
```diff
@@ -225,18 +225,30 @@ objects:
       to:
         kind: Service
         name: ${NAME}
-  - apiVersion: autoscaling/v1
+  - apiVersion: autoscaling/v2
```
Comment (dheepak-aot) [inline devops/openshift/api-deploy.yml:228]: 👍

File: devops/openshift/api-deploy.yml [hunk 2]
```diff
@@ -225,18 +225,30 @@ objects:
...
       minReplicas:  "${{REPLICAS}}"
       maxReplicas: 10
-      targetCPUUtilizationPercentage: 80
+      metrics:
+          - type: Resource
+            resource:
+              name: cpu
+              target:
+                type: AverageValue
+                averageValue: '0.18'
```
Comment (dheepak-aot) [inline devops/openshift/api-deploy.yml:245]: Asking for my clarification, how did we arrive at this average value?

File: devops/openshift/api-deploy.yml [hunk 3]
```diff
@@ -225,18 +225,30 @@ objects:
...
+          - type: Resource
+            resource:
+              name: memory
+              target:
+                type: AverageValue
+                averageValue: "230M"
```
Comment (dheepak-aot) [inline devops/openshift/api-deploy.yml:251]: Same here.

File: devops/openshift/forms-deploy.yml [hunk 4]
```diff
@@ -170,7 +170,19 @@ objects:
...
+              name: cpu
+              target:
+                type: AverageValue
+                averageValue: '0.8'
+          - type: Resource
+            resource:
+              name: memory
+              target:
+                type: AverageValue
+                averageValue: "1Gi"
```
Comment (dheepak-aot) [inline devops/openshift/forms-deploy.yml:185]: I see higher numbers for form.io server than API for autoscaling, is this with consideration that form.io server is going to have lighter load than API?

File: devops/openshift/api-deploy.yml (hunk 2 above)
Comment (guru-aot) [reply devops/openshift/api-deploy.yml:245]: 90% of the CPU limit

File: devops/openshift/forms-deploy.yml (hunk 4 above)
Comment (guru-aot) [reply devops/openshift/forms-deploy.yml:185]: No, the averageValue is a percentage of the CPU and memory limits and here for formio, we have used the 50% as we used in the previous load testing iterations, this inturn helps for pods scaling without many requests failing.

File: devops/openshift/api-deploy.yml (hunk 3 above)
Comment (guru-aot) [reply devops/openshift/api-deploy.yml:251]: 90% of the Memory limit

File: devops/openshift/forms-deploy.yml (hunk 4 above)
Comment (dheepak-aot) [reply devops/openshift/forms-deploy.yml:185]: Got it. Thanks for explaining. If 50% then cpu must be 0.75 ?

**Review by dheepak-aot — APPROVED** [review]
Great job 👍

File: devops/openshift/forms-deploy.yml (hunk 4 above)
Comment (guru-aot) [reply devops/openshift/forms-deploy.yml:185]: Yeah you are right, i did increased a bit and tested around so gave those value.

**Review by andrepestana-aot — APPROVED** [review]
Looks good! Good job!

File: devops/openshift/api-deploy.yml
Comment (andrewsignori-aot) [inline devops/openshift/api-deploy.yml:1]: Please ensure that this and all other documents are formatted. There are some minor differences. Please also ensure that VS Code is applying the format when saving 😉 
image

**Review by andrewsignori-aot — COMMENTED** [review]
Great work. Just one really minor comment.
Let's talk about extending or having a second ticket to implement the `scaleDown` behavior if needed.
Besides that, once deployed I would like to see the API autoscaling load test 😉

**Review by ann-aot — APPROVED** [review]
👍 Thanks for the details

**Pull request merged**